components/ai_chat/resources/page/components/input_box/editable.test.tsx

```diff
@@ -0,0 +1,339 @@
+          'Hi ',
+          createSkillNode('1', 'Alice'),
+          ', ',
+          createSkillNode('2', 'Bob'),
+          '!',
+        ]}
+        onContentChange={jest.fn()}
+      />,
+    )
+    const editable = container.querySelector('[data-editor]')!
+    expect(editable).toHaveTextContent('Hi Alice, Bob!')
+    expect(editable.querySelectorAll('span[data-type="skill"]')).toHaveLength(2)
+  })
+
+  it('preserves cursor position when typing (does not jump to start)', async () => {
+    const content: Content = ['Hello world']
+    const onContentChange = jest.fn()
+
+    const { container } = render(
+      <Editable
+        content={content}
+        onContentChange={onContentChange}
+      />,
+    )
+    const editable = container.querySelector('[data-editor]')!
+
+    // Position cursor in the middle of the text
+    const textNode = editable.firstChild as Text
+    const selection = window.getSelection()!
+    const range = document.createRange()
+    range.setStart(textNode, 6) // Position after "Hello "
+    range.setEnd(textNode, 6)
+    selection.removeAllRanges()
+    selection.addRange(range)
+
+    // Verify initial cursor position
+    expect(selection.focusOffset).toBe(6)
+
+    // Simulate typing by updating the text content and manually updating the cursor
+    editable.textContent = 'Hello Xworld'
+
+    // Manually set cursor position after the inserted character
+    const newTextNode = editable.firstChild as Text
+    const newRange = document.createRange()
+    newRange.setStart(newTextNode, 7) // After "Hello X"
+    newRange.setEnd(newTextNode, 7)
+    selection.removeAllRanges()
+    selection.addRange(newRange)
+
+    // Verify cursor is preserved (not at start)
+    const finalSelection = window.getSelection()!
+    expect(finalSelection.focusOffset).toBeGreaterThan(0)
+    expect(finalSelection.focusOffset).toBe(7) // After "Hello X"
+  })
+
+  it('onContentChange parses mixed content correctly from DOM', async () => {
+    const initialContent: Content = ['Hello']
+    const onContentChange = jest.fn()
+
+    const { container } = render(
+      <Editable
+        content={initialContent}
+        onContentChange={onContentChange}
+      />,
+    )
+    const editable = container.querySelector('[data-editor]')!
+
+    // Manually create DOM structure with mixed content
+    editable.innerHTML = ''
+
+    // Add text node
+    const textNode1 = document.createTextNode('Hello ')
+    editable.appendChild(textNode1)
+
+    // Add skill span
+    const span = document.createElement('span')
+    span.dataset.type = 'skill'
+    span.dataset.id = 'user-1'
+    span.dataset.text = 'Alice'
+    span.textContent = 'Alice'
+    span.contentEditable = 'false'
+    editable.appendChild(span)
+
+    // Add more text
+    const textNode2 = document.createTextNode(' welcome!')
+    editable.appendChild(textNode2)
+
+    // Trigger DOM mutation
+    triggerDOMMutation(editable)
+
+    await waitFor(() => {
```

Comment:
Would wrapping `triggerDOMMutation` in `act` help avoid `waitFor` here, considering it's meant to apply pending react updates